packaging/linux/deb/debroot.go

```diff
@@ -205,30 +217,123 @@ func fixupArtifactPerms(spec *dalec.Spec, target string, cfg *SourcePkgConfig) [
 	writeScriptHeader(buf, cfg)
 
 	basePath := filepath.Join("debian", spec.Name)
-
 	artifacts := spec.GetArtifacts(target)
-	if artifacts.Directories == nil {
-		return nil
-	}
 
-	sorted := dalec.SortMapKeys(artifacts.Directories.GetConfig())
-	for _, name := range sorted {
-		cfg := artifacts.Directories.Config[name]
+	checkAndWritePerms := func(key string, cfg dalec.ArtifactConfig, dir string) {
+		resolvedName := cfg.ResolveName(key)
+		p := filepath.Join(basePath, dir, resolvedName)
+
 		if cfg.Mode.Perm() != 0 {
-			p := filepath.Join(basePath, "etc", name)
 			fmt.Fprintf(buf, "chmod %o %q\n", cfg.Mode.Perm(), p)
+			return
+		}
+
+		// Debian does not keep original permissions for files, so we check if artifact matches a source name
+		// and if so, we apply the source permissions for inline sources.
+		srcMatch := strings.SplitN(key, "/", 2)
+		srcKey := srcMatch[0]
+		subpath := ""
+		if len(srcMatch) > 1 {
+			subpath = srcMatch[1]
+		}
+
+		src, ok := spec.Sources[srcKey]
+		if !ok || src.Inline == nil {
+			return
+		}
+
+		if src.Inline.File != nil && src.Inline.File.Permissions.Perm() != 0 {
+			fmt.Fprintf(buf, "chmod %o %q\n", src.Inline.File.Permissions.Perm(), p)
+			return
+		}
+
+		if src.Inline.Dir == nil {
+			return
+		}
+		if subpath == "" {
+			if src.Inline.Dir.Permissions.Perm() != 0 {
+				fmt.Fprintf(buf, "chmod %o %q\n", src.Inline.Dir.Permissions.Perm(), p)
+			}
+			return
+		}
+		if f, ok := src.Inline.Dir.Files[subpath]; ok && f.Permissions.Perm() != 0 {
+			fmt.Fprintf(buf, "chmod %o %q\n", f.Permissions.Perm(), p)
 		}
 	}
 
-	sorted = dalec.SortMapKeys(artifacts.Directories.GetState())
-	for _, name := range sorted {
-		cfg := artifacts.Directories.State[name]
-		if cfg.Mode.Perm() != 0 {
-			p := filepath.Join(basePath, "var/lib", name)
-			fmt.Fprintf(buf, "chmod %o %q\n", cfg.Mode.Perm(), p)
+	if artifacts.Binaries != nil {
+		for key, cfg := range artifacts.Binaries {
```

Comment:
In all of these cases we need to sort by key otherwise we'll end up busting the build cache potentially on every run due to randomness of maps.